x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java

```diff
@@ -83,47 +97,95 @@
+
+        if (concreteApplicationNames != null && concreteApplicationNames.size() == 0) {
+            logger.debug("returning empty application privileges as application names result in empty list");
+            listener.onResponse(Collections.emptySet());
+        } else {
+            final Set<ApplicationPrivilegeDescriptor> cachedDescriptors = cachedDescriptorsForApplicationNames(
+                concreteApplicationNames != null ? concreteApplicationNames : applicationNamesCacheKey);
+            if (cachedDescriptors != null) {
+                logger.debug("All application privileges found in cache");
+                listener.onResponse(filterDescriptorsForPrivilegeNames(cachedDescriptors, names));
+            } else {
+                final long invalidationCounter = numInvalidation.get();
+                // Always fetch all privileges of an application for caching purpose
+                logger.debug("Fetching application privilege documents for: {}", applicationNamesCacheKey);
+                innerGetPrivileges(applicationNamesCacheKey, ActionListener.wrap(fetchedDescriptors -> {
+                    final Map<String, Set<ApplicationPrivilegeDescriptor>> mapOfFetchedDescriptors = fetchedDescriptors.stream()
+                        .collect(Collectors.groupingBy(ApplicationPrivilegeDescriptor::getApplication, Collectors.toUnmodifiableSet()));
+                    // Avoid caching potential stale results.
+                    // TODO: It is still possible that cache gets invalidated immediately after the if check
+                    if (invalidationCounter == numInvalidation.get()) {
+                        final Set<String> fetchedApplicationNames = Collections.unmodifiableSet(mapOfFetchedDescriptors.keySet());
+                        // Do not cache the names if expansion has no effect
+                        if (fetchedApplicationNames.equals(applicationNamesCacheKey) == false) {
+                            logger.debug("Caching application names query: {} = {}", applicationNamesCacheKey, fetchedApplicationNames);
+                            applicationNamesCache.put(applicationNamesCacheKey, fetchedApplicationNames);
+                        }
+                        for (Map.Entry<String, Set<ApplicationPrivilegeDescriptor>> entry : mapOfFetchedDescriptors.entrySet()) {
+                            logger.debug("Caching descriptors for application: {}", entry.getKey());
+                            descriptorsCache.put(entry.getKey(), entry.getValue());
+                        }
+                    }
+                    listener.onResponse(filterDescriptorsForPrivilegeNames(fetchedDescriptors, names));
+                }, listener::onFailure));
+            }
+        }
+    }
+
+    private void innerGetPrivileges(Collection<String> applications,
+        ActionListener<Collection<ApplicationPrivilegeDescriptor>> listener) {
+        assert Objects.requireNonNull(applications).isEmpty() == false;
```

Comment:
That's a weird mix of `assert` and `require`.
```suggestion
        assert applications != null && applications.size() > 0 : "Application names are required (found " + applications + ")";
```